clang/tools/scan-build-py/lib/libear/ear.c

```diff
@@ -1759,156 +1744,261 @@ static void bear_report_call(char const *fun, char const *const argv[]) {
+      ret = 1;
+    } else if (is_nvcc_or_ld == 1) {
+      // object is not given by -o. Need figure out the default output for cmd
+      // "gcc -c xx.c"
+      char *tmp = malloc(PATH_MAX);
+      if (tmp == NULL) {
+        perror("bear: malloc memory fail.");
+        exit(EXIT_FAILURE);
+      }
+      memset(tmp, '\0', PATH_MAX);
+      int idx = 0;
+      int c_option;
       char ofilename[PATH_MAX];
-      int olen = strlen(argv[it]);
       memset(ofilename, '\0', PATH_MAX);
-      if (olen >= PATH_MAX) {
-        perror("bear: filename length too long.");
-        pthread_mutex_unlock(&mutex);
-        exit(EXIT_FAILURE);
+      int parse_ret = 0;
+      for (size_t it = it_cp; it < argc; ++it) {
+        memcpy(tmp + idx, argv[it], strlen(argv[it]));
+        idx += strlen(argv[it]);
+        tmp[idx] = ' ';
+        idx++;
       }
-      strncpy(ofilename, argv[it], olen);
-      if (generate_file(ofilename, 0) != 0) {
+      parse_ret = parse_input_file(tmp, &c_option, ofilename);
+      if (parse_ret == 1 && c_option == 1) {
+        // change the suffix of the ofilename from .c .cpp => .o)
+        int olen = strlen(ofilename);
+        while (olen >= 0 && ofilename[olen] != '.') {
+          olen--;
+        }
+        if (olen == -1) {
+          olen = strlen(ofilename);
+          if (olen >= PATH_MAX - 3) {
+            perror("bear: filename length too long.");
+            exit(EXIT_FAILURE);
+          }
+          ofilename[olen] = '.';
+          ofilename[olen + 1] = 'o';
+          ofilename[olen + 2] = '\0';
+        } else {
+          if (olen > PATH_MAX - 3) {
+            perror("bear: filename length too long.");
+            exit(EXIT_FAILURE);
+          }
+          ofilename[olen + 1] = 'o';
+          ofilename[olen + 2] = '\0';
+        }
+
+        pthread_mutex_lock(&mutex);
+        if (generate_file(ofilename, 0) != 0) {
+          pthread_mutex_unlock(&mutex);
+          exit(EXIT_FAILURE);
+        }
         pthread_mutex_unlock(&mutex);
-        exit(EXIT_FAILURE);
       }
-      flag_optval = 0;
-      flag_object = 1;
+      free(tmp);
+      ret = 1;
     }
-    if (flag_object == 0) {
-      // here we need parse out the object file if -o option is used.
-      // find xxx in the -o xxx of the command, generate it.
-      int r = find_create_object(tail);
-      if (r == 0) {
-        flag_object = 1;
-      }
-      if (r == 1) {
-        flag_optval = 1;
-      }
+
+    // try to replace nvcc or clang++ with intercept-stub,
+    // e.g: "/path/to/clang++ -ccbin ... ",
+    //      "/bin/sh -c "/path/to"/bin/clang++ -ccbin ..."
+    const char *pos = find_intercept_compiler(argv[it_cp], compiler_idx);
+
+    int is_stub_need = 0;
+    is_stub_need = (pos != NULL);
+    if (is_stub_need) {
+      ret = 0; // intercept-stub should continue to run.
+
+      // intercept-stub is used to handle the compiler command like
+      // "/bin/sh -c clang++ -c `echo ./`hello.c", it changes the compiler
```

Comment:
please keep footprint as small as possible for a PR.

Reply:
Those codes are in the `else branch` in the code below, logic keeps same as old code
`
  if (is_tool_existence) {
    for (size_t it = 0; it < argc; ++it) {
      fprintf(fd, "%s%c", argv[it], US);
    }
  } else {
    ...
}
`